File: basler_pyqt6/core/video_recorder.py
"""
視頻錄製器 - PyQt6 版本
簡化版本的工業相機錄製功能
"""

import cv2
import numpy as np
import logging
from pathlib import Path
from datetime import datetime
from typing import Optional

logger = logging.getLogger(__name__)


class VideoRecorder:
    """視頻錄製器"""

    def __init__(self, output_dir: str = "recordings"):
        """初始化錄製器

        Args:
            output_dir: 錄製文件輸出目錄
        """
        self.output_path = Path(output_dir)
        self.output_path.mkdir(exist_ok=True)

        self.is_recording = False
        self.video_writer: Optional[cv2.VideoWriter] = None
        self.current_filename = None
        self.frames_recorded = 0
        self.recording_start_time = None

        # 錄製參數
        self.fps = 30.0  # 預設FPS
        self.codec_name = None

        logger.info(f"✅ 視頻錄製器初始化完成，輸出目錄: {self.output_path}")

    def start_recording(
        self,
        frame_size: tuple = (640, 480),
        fps: float = 30.0,
        filename: str = None
    ) -> bool:
        """開始錄製

        Args:
            frame_size: 幀尺寸 (width, height)
            fps: 錄製幀率
            filename: 自定義文件名（不含副檔名）

        Returns:
            bool: 是否成功開始錄製
        """
        if self.is_recording:
            logger.warning("錄製已在進行中")
            return False

        try:
            # 生成文件名
            if filename is None:
                timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
                filename = f"recording_{timestamp}"

            self.current_filename = filename
            self.fps = fps

            # 嘗試不同的編碼器
            codecs_to_try = [
                ('mp4v', cv2.VideoWriter_fourcc(*'mp4v'), '.mp4'),  # 高品質
                ('MJPG', cv2.VideoWriter_fourcc(*'MJPG'), '.avi'),  # 穩定
                ('XVID', cv2.VideoWriter_fourcc(*'XVID'), '.avi')   # 備用
            ]

            for codec_name, codec_fourcc, extension in codecs_to_try:
                try:
                    filepath = self.output_path / f"{filename}{extension}"

                    self.video_writer = cv2.VideoWriter(
                        str(filepath),
                        codec_fourcc,
                        fps,
                        frame_size
                    )

                    if self.video_writer.isOpened():
                        self.codec_name = codec_name
                        logger.info(f"✅ 使用 {codec_name} 編碼器")
                        logger.info(f"📊 錄製參數: {frame_size[0]}x{frame_size[1]} @ {fps:.1f}fps")
                        logger.info(f"📁 錄製文件: {filepath}")
                        break
                    else:
                        self.video_writer = None

                except Exception as e:
                    logger.warning(f"⚠️ {codec_name} 編碼器失敗: {str(e)}")
                    self.video_writer = None

            if not self.video_writer or not self.video_writer.isOpened():
                logger.error("❌ 所有編碼器都失敗")
                return False

            self.is_recording = True
            self.frames_recorded = 0
            self.recording_start_time = datetime.now()

            logger.info(f"🎬 開始錄製: {filename}")
            return True

        except Exception as e:
            logger.error(f"錄製啟動失敗: {str(e)}")
            return False

    def write_frame(self, frame: np.ndarray) -> bool:
        """寫入一幀

        Args:
            frame: 要寫入的幀

        Returns:
            bool: 是否成功寫入
        """
        if not self.is_recording or self.video_writer is None:
            return False

        try:
            self.video_writer.write(frame)
            self.frames_recorded += 1
            return True

        except Exception as e:
            logger.error(f"寫入幀失敗: {str(e)}")
            return False

    def stop_recording(self) -> dict:
        """停止錄製

        Returns:
            dict: 錄製信息
        """
        if not self.is_recording:
            return {}

        try:
            self.is_recording = False

            # 計算錄製時長
            if self.recording_start_time:
                duration = (datetime.now() - self.recording_start_time).total_seconds()
            else:
                duration = 0.0

            # 釋放寫入器
            if self.video_writer:
                self.video_writer.release()
                self.video_writer = None

            # 計算平均FPS
            average_fps = self.frames_recorded / duration if duration > 0 else 0

            recording_info = {
                'filename': self.current_filename,
                'frames_recorded': self.frames_recorded,
                'duration': duration,
                'average_fps': average_fps,
                'codec': self.codec_name
            }

            logger.info(f"✅ 錄製完成: {self.current_filename}")
            logger.info(f"📊 錄製統計: {self.frames_recorded} 幀, {duration:.2f} 秒")
            logger.info(f"🎯 平均幀率: {average_fps:.1f} fps")

            return recording_info

        except Exception as e:
            logger.error(f"停止錄製失敗: {str(e)}")
            return {}

    def get_recording_status(self) -> dict:
        """獲取錄製狀態

        Returns:
            dict: 狀態信息
        """
        duration = 0.0
        if self.is_recording and self.recording_start_time:
            duration = (datetime.now() - self.recording_start_time).total_seconds()

        return {
            'is_recording': self.is_recording,
            'filename': self.current_filename,
            'frames_recorded': self.frames_recorded,
            'duration': duration,
            'fps': self.fps
        }

    def cleanup(self):
        """清理資源"""
        if self.is_recording:
            self.stop_recording()

        if self.video_writer:
            self.video_writer.release()
            self.video_writer = None

        logger.info("✅ 錄製器資源已清理")

File: basler_cpp/src/core/video_recorder.cpp
#include "core/video_recorder.h"
#include <QDebug>
#include <QSize>

namespace basler {

VideoRecorder::VideoRecorder(const QString& outputDir, QObject* parent)
    : QObject(parent)
    , m_outputPath(outputDir)
{
    // 確保輸出目錄存在
    if (!m_outputPath.exists()) {
        m_outputPath.mkpath(".");
    }

    qDebug() << "[VideoRecorder] 初始化完成，輸出目錄:" << m_outputPath.absolutePath();
}

VideoRecorder::~VideoRecorder()
{
    cleanup();
}

double VideoRecorder::recordingDuration() const
{
    if (!m_isRecording.load() || !m_recordingStartTime.isValid()) {
        return 0.0;
    }
    return m_recordingStartTime.msecsTo(QDateTime::currentDateTime()) / 1000.0;
}

void VideoRecorder::setOutputDirectory(const QString& dir)
{
    m_outputPath = QDir(dir);
    if (!m_outputPath.exists()) {
        m_outputPath.mkpath(".");
    }
}

bool VideoRecorder::startRecording(const QSize& frameSize, double fps, const QString& filename)
{
    if (m_isRecording.load()) {
        qWarning() << "[VideoRecorder] 錄製已在進行中";
        return false;
    }

    // 生成文件名
    QString actualFilename = filename;
    if (actualFilename.isEmpty()) {
        QString timestamp = QDateTime::currentDateTime().toString("yyyyMMdd_HHmmss");
        actualFilename = QString("recording_%1").arg(timestamp);
    }
    m_currentFilename = actualFilename;
    m_fps = fps;

    // 嘗試不同的編碼器
    struct CodecInfo {
        QString name;
        int fourcc;
        QString extension;
    };

    std::vector<CodecInfo> codecs = {
        {"mp4v", cv::VideoWriter::fourcc('m', 'p', '4', 'v'), ".mp4"},
        {"MJPG", cv::VideoWriter::fourcc('M', 'J', 'P', 'G'), ".avi"},
        {"XVID", cv::VideoWriter::fourcc('X', 'V', 'I', 'D'), ".avi"}
    };

    for (const auto& codec : codecs) {
        if (tryCodec(codec.name, codec.fourcc, codec.extension, frameSize, fps)) {
            m_isRecording.store(true);
            m_framesRecorded.store(0);
            m_recordingStartTime = QDateTime::currentDateTime();

            emit recordingStarted(m_currentFilename);
            emit recordingStateChanged(true);

            qDebug() << "[VideoRecorder] 開始錄製:" << m_currentFilename;
            return true;
        }
    }

    emit recordingError("所有編碼器都失敗");
    return false;
}

bool VideoRecorder::tryCodec(const QString& codecName, int fourcc, const QString& extension,
                             const QSize& frameSize, double fps)
{
    try {
        QString filepath = m_outputPath.absoluteFilePath(m_currentFilename + extension);

        m_videoWriter = std::make_unique<cv::VideoWriter>(
            filepath.toStdString(),
            fourcc,
            fps,
            cv::Size(frameSize.width(), frameSize.height())
        );

        if (m_videoWriter->isOpened()) {
            m_codecName = codecName;
            m_currentFullPath = filepath;

            qDebug() << "[VideoRecorder] 使用" << codecName << "編碼器";
            qDebug() << "[VideoRecorder] 錄製參數:"
                     << frameSize.width() << "x" << frameSize.height()
                     << "@" << fps << "fps";
            qDebug() << "[VideoRecorder] 錄製文件:" << filepath;

            return true;
        }

        m_videoWriter.reset();
        return false;

    } catch (const std::exception& e) {
        qWarning() << "[VideoRecorder]" << codecName << "編碼器失敗:" << e.what();
        m_videoWriter.reset();
        return false;
    }
}

bool VideoRecorder::writeFrame(const cv::Mat& frame)
{
    if (!m_isRecording.load() || !m_videoWriter) {
        return false;
    }

    QMutexLocker locker(&m_writerMutex);

    try {
        m_videoWriter->write(frame);
        int frames = m_framesRecorded.fetch_add(1) + 1;
        emit frameWritten(frames);
        return true;

    } catch (const std::exception& e) {
        qWarning() << "[VideoRecorder] 寫入幀失敗:" << e.what();
        return false;
    }
}

RecordingInfo VideoRecorder::stopRecording()
{
    RecordingInfo info;

    if (!m_isRecording.load()) {
        return info;
    }

    m_isRecording.store(false);

    // 計算錄製時長
    double duration = 0.0;
    if (m_recordingStartTime.isValid()) {
        duration = m_recordingStartTime.msecsTo(QDateTime::currentDateTime()) / 1000.0;
    }

    // 釋放寫入器
    {
        QMutexLocker locker(&m_writerMutex);
        if (m_videoWriter) {
            m_videoWriter->release();
            m_videoWriter.reset();
        }
    }

    // 填充錄製信息
    int frames = m_framesRecorded.load();
    info.filename = m_currentFilename;
    info.fullPath = m_currentFullPath;
    info.framesRecorded = frames;
    info.duration = duration;
    info.averageFps = (duration > 0) ? (frames / duration) : 0.0;
    info.codec = m_codecName;

    emit recordingStopped(info);
    emit recordingStateChanged(false);

    qDebug() << "[VideoRecorder] 錄製完成:" << m_currentFilename;
    qDebug() << "[VideoRecorder] 錄製統計:" << frames << "幀," << duration << "秒";
    qDebug() << "[VideoRecorder] 平均幀率:" << info.averageFps << "fps";

    return info;
}

void VideoRecorder::cleanup()
{
    if (m_isRecording.load()) {
        stopRecording();
    }

    QMutexLocker locker(&m_writerMutex);
    if (m_videoWriter) {
        m_videoWriter->release();
        m_videoWriter.reset();
    }

    qDebug() << "[VideoRecorder] 資源已清理";
}

} // namespace basler
